Add localState ngrxDirective option

mucaho opened the pull request:
@MrWolfZ 
As discussed in issue #165 , this PR adds:
* The `ngrxSuppressGlobalActionEmit` input boolean to ngrx's directives, which controls whether to dispatch form actions to the global `ActionsSubject` or not, `false` by default of course to not change current semantics. I thought this was better wording than `createLocalFormState`, but I can adapt this of course.
* The `ngrxFormAction` output event emitter that **always** emits any form action. It does not hurt if the action is always emitted, regardless of `ngrxSuppressGlobalActionEmit`, right?
* Unit and e2e tests
* A representative example showcasing the benefit of using local state. Showcases how to handle state update on inter-dependent form state and additional local state used in the component. Showcases how to properly detect changes in lieu of input properties, but still being as efficient as possible.
* Does **not** include any documentation. To be frank, I do not know where to put and what to write what's not already showcased in example. Can change this of course, depending on your input.

I hope everything checks out, otherwise let me know.

(e: Improved upon the example and some wording, pushed again)

MrWolfZ commented:
Thanks a lot for this thorough PR. Overall it looks great. Here's a couple of comments:

> The `ngrxSuppressGlobalActionEmit` input boolean to ngrx's directives, which controls whether to dispatch form actions to the global `ActionsSubject` or not, false by default of course to not change current semantics. I thought this was better wording than `createLocalFormState`, but I can adapt this of course.

I like this design because it keeps the whole local form state concept in the directive without bleeding into the rest of **ngrx-forms**. At the same time I dislike its verbosity, because you need to pass the parameter to every single form control (which also makes it easy to accidentally forget adding it on a control). If there was an `isLocal` property on the form state then the directive could decide based on that, but this would require some changes to other areas. Thinking about it, there may also be a slightly different API design that at least takes care of the verbosity issue: instead of passing an additional flag there could be a new directive that uses `ngrxLocalFormControlState` (or `ngrxFormLocalControlState`) as a selector. This design still has the issue of it being possible to accidentally mix local and global form states though. What do you think?

> The `ngrxFormAction` output event emitter that always emits any form action. It does not hurt if the action is always emitted, regardless of `ngrxSuppressGlobalActionEmit`, right?

Yeah, its fine to always emit locally.

> A representative example showcasing the benefit of using local state. Showcases how to handle state update on inter-dependent form state and additional local state used in the component. Showcases how to properly detect changes in lieu of input properties, but still being as efficient as possible.

I feel that the example is maybe a bit too advanced to make the feature easy to understand. There should be at least one basic example that showcases purely how to handle only local form state without any interaction with effects etc, i.e. showing that this feature allows using **ngrx-forms** without ngrx. There could then be a second more advanced example for how to connect a local form with actions from the outside.

> Does not include any documentation. To be frank, I do not know where to put and what to write what's not already showcased in example. Can change this of course, depending on your input.

Yeah, a good example is always important. Even though the text documentation may not contain any extra details, it is still useful, since there are different people out there that learn through different things. Some may prefer looking at an example, others may prefer reading through documentation. For this feature we could probably just add a new markdown file `local-forms.md` to the `user-guide` section.

Overall, great work. Let's see if we can nail down the API design. While it's great that the PR is already so complete, in the future I would recommend to not start writing any documentation/examples before the design is finalized to save yourself some unnecessary refactoring/rewriting.

MrWolfZ commented on example-app/src/app/local-state/local-state.component.ts:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  ngOnDestroy(): void {
+    this.subscription.unsubscribe();
```
a tiny detail I like to keep in mind here is that `ngOnDestroy` can be called without `ngOnInit`, and therefore `this.subscription` can be undefined. You can either surround this with an `if` or initialize the variable upon declaration.

MrWolfZ commented on example-app/src/app/local-state/local-state.reducer.ts:
```diff
@@ -0,0 +1,74 @@
+const reducers = combineReducers<LocalState>({
+  manufacturers(manufacturers = [], a: Action) {
+    // update from loaded data
+    if (a instanceof SetManufacturersAction) {
```
personally I never use `instanceof` with actions, but instead compare the `type`. This is because it is not guaranteed the action will always be a class instance (e.g. if dispatched via the redux devtools).

MrWolfZ commented on src/control/directive.spec.ts:
```diff
@@ -560,6 +560,113 @@ describe(NgrxFormControlDirective.name, () => {
+      expect(directive.ngrxSuppressGlobalActionEmit).toBe(false);
+    });
+
+    describe('is disabled', () => {
```
aren't these tests redundant since they represent the existing behavior that is already tested?

MrWolfZ commented on src/control/directive.ts:
```diff
@@ -86,6 +88,10 @@ export class NgrxFormControlDirective<TStateValue extends FormControlValueTypes,
 
   @Input() ngrxValueConverter: NgrxValueConverter<TViewValue, TStateValue> = NgrxValueConverters.default<any>();
 
+  @Input() ngrxSuppressGlobalActionEmit: boolean = false;
```
minor thing but the type annotation is redundant. i'd remove it for consistency with the rest of the code

MrWolfZ commented on src/control/e2e.spec/select.spec.ts:
```diff
@@ -172,3 +172,70 @@ describe(NumberSelectComponent.name, () => {
   });
 });
+
+@Component({
```
the reason for `e2e.spec/select.spec.ts` is that there were some tricky interaction issues with angular itself. This is unrelated to the local form state. It's still a good idea to have an end to end test though, but I'd put it in a separate `local-state.spec.ts` file and maybe use just an `input` instead of the more complex `select`

mucaho replied:
Was not able to trigger a simple change event manually on an input element and have ngrx forms register it properly. Extracted the example into a separate spec file though.

MrWolfZ commented on src/group/directive.spec.ts:
```diff
@@ -46,4 +46,107 @@ describe(NgrxFormDirective.name, () => {
   });
+
+  describe('suppression of global action emit', () => {
+    it('be disabled by default', () => {
```
missing `should`?

MrWolfZ commented on src/group/directive.spec.ts:
```diff
@@ -46,4 +46,107 @@ describe(NgrxFormDirective.name, () => {
+      expect(directive.ngrxSuppressGlobalActionEmit).toBe(false);
+    });
+
+    describe('is disabled', () => {
```
same as for control directive, I think these tests are redundant

mucaho commented:
Thanks for the review, i will incorporate all the remarks once we settle on the API, as you said.

One thing I am not certain about: For now we have the `ngrxFormAction` output on both `ngrxFormState` and `ngrxFormControlState`. Should the output be called `ngrxFormControlAction` in the latter?

 > At the same time I dislike _[the `ngrxSuppressGlobalActionEmit` input boolean's]_  verbosity, because you need to pass the parameter to every single form control (which also makes it easy to accidentally forget adding it on a control). If there was an `isLocal` property on the form state then the directive could decide based on that, but this would require some changes to other areas. Thinking about it, there may also be a slightly different API design that at least takes care of the verbosity issue: instead of passing an additional flag there could be a new directive that uses `ngrxLocalFormControlState` (or `ngrxFormLocalControlState`) as a selector. This design still has the issue of it being possible to accidentally mix local and global form states though. What do you think?

* The `isLocal` form state property would be the least verbose one, and once you have set it, you would not have to think about using the proper input / directive for all your form controls. While this sounds nice I'm not a big fan of mixing form state with configuration options.
*  `ngrxLocalFormControlState` has the same issue of being accidentally mixed up or forgotten about like `ngrxSuppressGlobalActionEmit`. However, it's less verbose as you eliminate the need for this additional input.

In my opinion another set of directives (`ngrxLocalFormControlState` & ``ngrxLocalFormState`) would be the best option.

If you agree, I would proceed with this approach. However, in this case the question remains how to go about the implementation. As far as I can tell, there are two options:
* make `ngrxLocalFormControlState` directive extend `ngrxFormControlState` directive, passing the former's output `EventEmitter` in the super constructor as the `ActionsSubject`. A bit hacky since `EventEmitter` being a subject is an implementation detail, which theoretically could change in future releases.
* rename current `ngrxFormControlState` directive to `ngrxLocalFormControlState`. Remove `ActionsSubject` from constructor and only emit to the local `ngrxFormAction` output.
Make a new directive called `ngrxFormControlState` which has the same constructor parameters plus the `ActionsSubject` and overrides the `dispatchAction` method by also emitting to this `ActionsSubject`. (And hope for the best that angular inherits all decorators properly).

MrWolfZ commented:
> One thing I am not certain about: For now we have the ngrxFormAction output on both ngrxFormState and ngrxFormControlState. Should the output be called ngrxFormControlAction in the latter?

Hmm, I am honestly divided. In the grand scheme of things this is probably just a minor detail though, so any option is fine. We could also just call it `ngrxFormsAction` to indicate these are actions from this library. Choose whatever feels right to you.

> In my opinion another set of directives (`ngrxLocalFormControlState` & `ngrxLocalFormState`) would be the best option.

Yeah, let's go with this design.

> If you agree, I would proceed with this approach. However, in this case the question remains how to go about the implementation. As far as I can tell, there are two options:

I was thinking about this as well. I would go for the inheritance approach. You can simply make the `ActionsSubject` constructor parameter `@Optional()` and then provide `null` as a value to the super class constructor in the new directive. In addition you would keep the `dispatchAction` method and just override it in the subclass to emit on the local `EventEmitter` instead of the `ActionsSubject`. This should work for both directives.

mucaho commented:
I encountered some unexpected roadblocks while trying to implement the directive `ngrxLocalFormControlState` as a subclass of `ngrxFormControlState`:
* My e2e test kept failing, much to my surprise. Reason is that the new directive `ngrxLocalFormControlState` can not find any `@Self`-scoped `FormViewAdapter`s. After some investigation I found out that these are added by the same setter property `ngrxFormControlState` in the various adapters. While adding another selector and input setter would be possible for in-house viewAdapters that same can not be said for external adapters that exist already and rely on this naming convention. e.g. for `NgrxRadioViewAdapter`:
```ts
@Directive({
  selector: 'input[type=radio][ngrxFormControlState]',
  // ...
})
export class NgrxRadioViewAdapter implements FormViewAdapter, OnInit, AfterViewInit {
  @Input() set ngrxFormControlState(value: FormControlState<any>) { /*...*/ }
}
```
* The documentation section on [user-defined properties](https://ngrx-forms.readthedocs.io/en/master/user-guide/form-controls/#user-defined-properties) uses the same, aforementioned naming convention. Same problem again.
* Same problem for `CustomErrorStateMatcherDirective`, although I don't know for what that's used.

Here the implementation for reference:
```ts
@Directive({
  // tslint:disable-next-line:directive-selector
  selector: '[ngrxLocalFormControlState]',
})
export class NgrxLocalFormControlDirective<TStateValue extends FormControlValueTypes, TViewValue = TStateValue>
  extends NgrxFormControlDirective<TStateValue, TViewValue> {

  @Input() set ngrxLocalFormControlState(state: FormControlState<TStateValue>) {
    this.ngrxFormControlState = state;
  }

  @Output() ngrxFormsAction = new EventEmitter<Actions<TStateValue>>();

  constructor(
    el: ElementRef,
    @Optional() @Inject(DOCUMENT) dom: Document | null,
    @Self() @Optional() @Inject(NGRX_FORM_VIEW_ADAPTER) viewAdapters: FormViewAdapter[],
    @Self() @Optional() @Inject(NG_VALUE_ACCESSOR) valueAccessors: ControlValueAccessor[],
  ) {
    super(el, dom, null as any as ActionsSubject, viewAdapters, valueAccessors);
  }

  protected dispatchAction(action: Actions<TStateValue>) {
    this.ngrxFormsAction.emit(action);
  };

}
```

-----

If you know of a good way of introducing another directive, while still being backward-compatible let me know. A cool solution would be to apply the new directive on all elements that have the output binding too, something like `selector: [ngrxFormControlState] (ngrxFormsAction)`, but that does not work of course.

Otherwise, maybe we should go ahead with the original approach. Looking forward to hear your thoughts.

mucaho commented:
Changes from previous version:
* Added the local directives. as mentioned in previous comment they are added if you bind an event listener to `(ngrxFormsAction)`, otherwise the default directives are used.
* Add an introductory local state example.
* Added documentation.

Codecov is failing because I did not test the branches where the `@Optional`ly injected `ActionsSubject` is `null`. Should I throw an error in this case? It should never happen, as I override the `dispatchMethod` in that case when it's null.

MrWolfZ commented:
> Scratch everything I wrote before, I can make the new directive's selector be `[ngrxFormControlState][ngrxFormsAction]` and change the old one's to `[ngrxFormControlState]:not([ngrxFormsAction])` and everything looks to be working properly.

Yeah, that's a pretty cool approach. For reference in the future, I believe you can also have multiple selectors for a component, e.g `[ngrxFormControlState],[ngrxLocalFormControlState]`. However, I believe your solution is much more elegant since it is much less invasive than using a different attribute as a selector. I'll take a look at the code.

MrWolfZ commented on docs/user-guide/local-form-state.md:
```diff
@@ -0,0 +1,51 @@
+        (ngrxFormsAction)="handleFormAction($event)">
+    </form>
+  `,
+  changeDetection: ChangeDetectionStrategy.OnPush,
```
I'd remove this from the example.

MrWolfZ commented on docs/user-guide/local-form-state.md:
```diff
@@ -0,0 +1,51 @@
+Maintaining a local form state has its merits:
+
+* The root or feature store is not populated with **temporary form data** which perishes after the component is destroyed.
+Form data is often used for creating and or updating an entity and becomes obsolete after an update request is sent to the server. After the server responds, you can update your list or table of entities with the new or updated, proper entity data.
```
typo in `creating and or updating`? If you want both, separate them with a slash, i.e. `creating and/or updating`.

MrWolfZ commented on example-app/src/app/local-state-advanced/local-state-advanced.component.html:
```diff
@@ -0,0 +1,52 @@
+<ngf-form-example exampleName="Local state: Advanced"
+                  [formState]="localState"
+                  >
+  Advanced example that shows how you can disable emitting any form actions to the global store via the builtin ActionSubject.
```
typo `ActionSubject` -> `ActionsSubject`

MrWolfZ commented on example-app/src/app/local-state-advanced/local-state-advanced.component.html:
```diff
@@ -0,0 +1,52 @@
+  Instead, you can listen for actions using the directive's output EventEmitter and route them to the component.
+  <br />
+  The component holds, updates and maintains a local state, which does not taint the global store and
+  is automatically lost once the component is destroyed.
```
`lost` sounds quite negative, I would use `cleaned up` or `removed`

MrWolfZ commented on example-app/src/app/local-state-advanced/local-state-advanced.component.ts:
```diff
@@ -0,0 +1,55 @@
+
+import { reducer, LocalState, GetManufacturersAction, INITIAL_LOCAL_STATE, FORM_ID } from './local-state-advanced.reducer';
+
+
```
remove this empty line

MrWolfZ commented on example-app/src/app/local-state-advanced/local-state-advanced.component.ts:
```diff
@@ -0,0 +1,55 @@
+})
+export class LocalStateAdvancedComponent implements OnInit, OnDestroy {
+
+  public localState: LocalState = INITIAL_LOCAL_STATE;
```
`public` and type declaration are redundant

MrWolfZ commented on example-app/src/app/local-state-advanced/local-state-advanced.component.ts:
```diff
@@ -0,0 +1,55 @@
+    this.updateState(action);
+
+    // trigger loading of new manufacturers list in effect
+    if (action.type === SetValueAction.TYPE && action.controlId === `${FORM_ID}.countryCode`) {
```
Instead of `${FORM_ID}.countryCode` which is not refactoring safe, I would just use `this.localState.formState.controls.CountryCode.id`, even if it is longer

MrWolfZ commented on example-app/src/app/local-state-advanced/local-state-advanced.reducer.ts:
```diff
@@ -0,0 +1,74 @@
+import { createFormGroupState, formGroupReducer, FormGroupState, updateGroup, setValue } from 'ngrx-forms';
+
+export class GetManufacturersAction implements Action {
+  static readonly TYPE = 'localState/GET_MANUFACTURERS';
```
change type prefix to `localStateAdvanced`

MrWolfZ commented on example-app/src/app/local-state-introduction/local-state-introduction.component.html:
```diff
@@ -0,0 +1,26 @@
+<ngf-form-example exampleName="Local state: Introduction"
+                  [formState]="formState"
+                  >
+  Introductory example that shows how you can disable emitting any form actions to the global store via the builtin ActionSubject.
```
typo `ActionSubject` -> `ActionsSubject`

MrWolfZ commented on example-app/src/app/local-state-introduction/local-state-introduction.component.html:
```diff
@@ -0,0 +1,26 @@
+  Instead, you can listen for actions using the directive's output EventEmitter and route them to the component.
+  <br />
+  The component holds, updates and maintains a local form state, which does not taint the global store and
+  is automatically lost once the component is destroyed.
```
`cleaned up` or `removed` instead of `lost`

MrWolfZ commented on example-app/src/app/local-state-introduction/local-state-introduction.component.ts:
```diff
@@ -0,0 +1,20 @@
+  selector: 'ngf-local-state-introduction',
+  templateUrl: './local-state-introduction.component.html',
+  styleUrls: ['./local-state-introduction.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
```
remove `changeDetection`

MrWolfZ commented on example-app/src/app/local-state-introduction/local-state-introduction.component.ts:
```diff
@@ -0,0 +1,20 @@
+})
+export class LocalStateIntroductionComponent {
+
+  public formState: FormGroupState<FormValue> = INITIAL_FORM_STATE;
```
`public` and type declaration are redundant

MrWolfZ commented on example-app/src/app/local-state-introduction/local-state-introduction.reducer.ts:
```diff
@@ -0,0 +1,17 @@
+import { Action } from '@ngrx/store';
+import { createFormGroupState, formGroupReducer, FormGroupState } from 'ngrx-forms';
+
+
```
remove extra blank line

MrWolfZ commented on src/control/directive.ts:
```diff
@@ -53,7 +53,7 @@ class ControlValueAccessorAdapter implements FormViewAdapter {
 @Directive({
   // tslint:disable-next-line:directive-selector
-  selector: '[ngrxFormControlState]',
+  selector: ':not([ngrxFormsAction])[ngrxFormControlState]',
```
i'd prefer `[ngrxFormControlState]:not([ngrxFormsAction])`

mucaho replied:
See https://github.com/MrWolfZ/ngrx-forms/pull/166#discussion_r349931342

MrWolfZ commented on src/control/directive.ts:
```diff
@@ -112,7 +112,7 @@ export class NgrxFormControlDirective<TStateValue extends FormControlValueTypes,
     // there is no reference to the Document type to support non-browser platforms
     @Optional() @Inject(DOCUMENT) private dom: Document | null,
-    private actionsSubject: ActionsSubject,
+    @Optional() private actionsSubject: ActionsSubject,
```
if it is optional, then you should type it as `ActionsSubject | null` since Angular provides `null` for non-existing optional dependencies. This will ensure you are not accidentally accessing the subject anywhere unsafely

mucaho replied:
That is not possible due to how the Angular DI mechanism works. If you extend it to union type `ActionsSubject | null` it will complain that it can not find the required `InjectionToken`. I have introduced a private field with the proper type declaration and just set it the parameter to the field.

MrWolfZ replied:
Hmm, I believe it would work if you specify `@Inject(ActionsSubject)` but I am honestly not 100% sure. I'll look into this myself after merging.

mucaho replied:
Ah, I see, good to know that it can be done more elegantly with this.

MrWolfZ commented on src/control/e2e.spec/local-state.spec.ts:
```diff
@@ -0,0 +1,79 @@
+import { NgrxFormsModule } from '../../module';
+import { createFormControlState, FormControlState } from '../../state';
+
+
```
remove extra blank line

MrWolfZ commented on src/control/e2e.spec/local-state.spec.ts:
```diff
@@ -0,0 +1,79 @@
+    element = nativeElement.querySelector('select')!;
+  });
+
+  it(`should not trigger a ${MarkAsDirtyAction.name}to the global store when an option is selected`, done => {
```
missing space before `to`

MrWolfZ commented on src/control/local-state-directive.ts:
```diff
@@ -0,0 +1,42 @@
+import { NgrxFormControlDirective, Document } from './directive';
+import { Actions } from '../actions';
+
+
```
remove extra blank line

MrWolfZ commented on src/control/local-state-directive.ts:
```diff
@@ -0,0 +1,42 @@
+    @Self() @Optional() @Inject(NGRX_FORM_VIEW_ADAPTER) viewAdapters: FormViewAdapter[],
+    @Self() @Optional() @Inject(NG_VALUE_ACCESSOR) valueAccessors: ControlValueAccessor[],
+  ) {
+    super(el, dom, null as any as ActionsSubject, viewAdapters, valueAccessors);
```
you can remove the casts when the base class constructor accepts `null`

mucaho replied:
See https://github.com/MrWolfZ/ngrx-forms/pull/166#discussion_r349931342

MrWolfZ commented on src/control/local-state-directive.ts:
```diff
@@ -0,0 +1,42 @@
+  protected dispatchAction(action: Actions<TStateValue>) {
+    this.ngrxFormsAction.emit(action);
+  };
+
```
remove extra blank line

MrWolfZ commented on src/group/directive.ts:
```diff
@@ -11,13 +11,19 @@ interface CustomEvent extends Event { }
 @Directive({
   // tslint:disable-next-line:directive-selector
-  selector: 'form[ngrxFormState]',
+  selector: 'form:not([ngrxFormsAction])[ngrxFormState]',
```
i'd prefer `form[ngrxFormState]:not([ngrxFormsAction])`

mucaho replied:
Tried that initially, but the selector won't work this way.

MrWolfZ commented on src/group/directive.ts:
```diff
@@ -11,13 +11,19 @@ interface CustomEvent extends Event { }
   @Input('ngrxFormState') state: FormGroupState<TValue>;
 
-  constructor(private actionsSubject: ActionsSubject) { }
+  constructor(@Optional() private actionsSubject: ActionsSubject) { }
```
type as `ActionsSubject | null`

mucaho replied:
See https://github.com/MrWolfZ/ngrx-forms/pull/166#discussion_r349931342

MrWolfZ commented on src/group/local-state-directive.ts:
```diff
@@ -0,0 +1,24 @@
+  @Output() ngrxFormsAction = new EventEmitter<Actions<TValue>>();
+
+  constructor() {
+    super(null as any as ActionsSubject);
```
can remove cast when base constructor accepts `null`

mucaho replied:
See https://github.com/MrWolfZ/ngrx-forms/pull/166#discussion_r349931342

mucaho commented:
Alright, applied all proposed changes, or left responses where it's not possible to do so.
I hope we are good to go now.

MrWolfZ commented:
Yup, we are good.

MrWolfZ commented:
And it's published as version [6.1.0](https://github.com/MrWolfZ/ngrx-forms/releases/tag/v6.1.0).

mucaho commented:
Thanks for all your input, I am glad we managed to find a nice API and provide it with examples & documentation!

I see you have done some linting afterwards, thanks!
As you have mentioned already before, I think you can add `tslint` to the `npm test` script, so the CI check fails if it's not properly formatted. This would remove the need for the manual chore work of fixing formatting, as you can run `tslint --fix` to resolve most of the problems.
